packages/libraries/core/src/client/persisted-documents.ts

```diff
@@ -154,33 +92,97 @@ export function createPersistedDocuments(
     return circuitBreaker;
   });
 
-  /** Batch load a persisted documents */
-  function loadPersistedDocument(documentId: string) {
-    const validationError = validateDocumentId(documentId);
-    if (validationError) {
-      // Return a promise that will be rejected with a proper error
-      return Promise.reject(createValidationError(documentId, validationError.error));
+  // Attempt to get document from L2 cache, returns: { hit: true, value: string | null } or { hit: false }
+  async function getFromLayer2Cache(
+    documentId: string,
+  ): Promise<{ hit: true; value: string | null } | { hit: false }> {
+    if (!layer2Cache) {
+      return { hit: false };
+    }
+
+    let cached: string | typeof PERSISTED_DOCUMENT_NOT_FOUND | null;
+    try {
+      cached = await layer2Cache.get(documentId);
+    } catch (error) {
+      // L2 cache failure should not break the request
+      config.logger.warn('L2 cache get failed for document %s: %O', documentId, error);
+      return { hit: false };
     }
 
-    const document = persistedDocumentsCache.get(documentId);
-    if (document) {
-      return document;
+    if (cached === null) {
+      // Cache miss
+      return { hit: false };
     }
 
+    if (cached === PERSISTED_DOCUMENT_NOT_FOUND) {
+      // Negative cache hit, document was previously not found
+      config.logger.debug('L2 cache negative hit for document %s', documentId);
+      return { hit: true, value: null };
+    }
+
+    // Cache hit with document
+    config.logger.debug('L2 cache hit for document %s', documentId);
+    return { hit: true, value: cached };
+  }
+
+  // store document in L2 cache (fire-and-forget, non-blocking)
+  function setInLayer2Cache(documentId: string, value: string | null): void {
+    if (!layer2Cache?.set) {
+      return;
+    }
+
+    // Skip negative caching if TTL is 0
+    if (value === null && layer2NotFoundTtlSeconds === 0) {
+      return;
+    }
+
+    const cacheValue = value === null ? PERSISTED_DOCUMENT_NOT_FOUND : value;
+    const ttl = value === null ? layer2NotFoundTtlSeconds : layer2TtlSeconds;
+
+    // Fire-and-forget. don't await, don't block
+    const setPromise = layer2Cache.set(documentId, cacheValue, ttl ? { ttl } : undefined);
+    if (setPromise && typeof setPromise.catch === 'function') {
+      setPromise.catch(error => {
```

Comment:
Floating promises are tricky in some environments(especially serverless ones). Those environments don't wait for event loop as Node does, and in this case we use `waitUntil`(from the context, available in Yoga so in Hive GW) to keep the worker open even if the response is returned to let the promise resolved, or explicit resource management to prevent a shutdown without resolving the promise/job.
https://the-guild.dev/graphql/yoga-server/docs/features/explicit-resource-management#background-jobs